cpp/src/arrow/util/int-util.cc

```diff
@@ -402,5 +402,45 @@ void DowncastUInts(const uint64_t* source, uint64_t* dest, int64_t length) {
   memcpy(dest, source, length * sizeof(int64_t));
 }
 
+template <typename InputInt, typename OutputInt>
+void TransposeInts(const InputInt* src, OutputInt* dest, int64_t length,
+                   const int32_t* transpose_map) {
+  while (length >= 4) {
+    dest[0] = static_cast<OutputInt>(transpose_map[src[0]]);
+    dest[1] = static_cast<OutputInt>(transpose_map[src[1]]);
+    dest[2] = static_cast<OutputInt>(transpose_map[src[2]]);
```

Comment:
Aren't you concerned by truncation due to incompatible size, maybe a static_assert somewhere?

Reply:
What would we `static_assert`? Indeed it's up to the caller to ensure the input is ok (it is by construction in case of dictionary unification).

Reply:
sizeof(in_type) <= sizeof(out_type)
is_signed(in_type) == is_signed(out_type)

Reply:
Ah, but the converse is supported as well.